File: pset6/credit/credit.py
import re

def main():
    """
    Applying Luhn's Algo
    Multiplying every other digit (from the right) by 2
    and finding the sum. Then adding the sum of the digits
    not multiplied by 2
    Returns Credit Card Brand if valid Card and
    INVALID otherwise
    """

    cardNum = input("Number: ")

    if len(cardNum) % 2 == 0:
        products = list(range(0 , len(cardNum), 2))
        sums = list(range(1, len(cardNum), 2))
    else:
        products = list(range(1, len(cardNum), 2))
        sums = list(range(0, len(cardNum), 2))

    # Storing the products in a list
    ls = []
    for i in products:
        # Getting the product
        num = int(cardNum[i]) * 2
        # Converting to a string
        s = str(num)
        # Appending the digits to ls
        for c in s:
            ls.append(int(c))

    # Intializing the luhn_val with the sum of the products
    luhn_val = sum(ls)

    for i in sums:
        num = int(cardNum[i])
        luhn_val += num

    if luhn_val % 10 != 0 or len(cardNum) not in [13, 15, 16]:
        print("INVALID")
        return

    if re.search("^51|^52|^53|^54|^55", cardNum):
        print("MASTERCARD")
    elif re.search("^34|^37", cardNum):
        print("AMEX")
    elif re.search("^4", cardNum):
        print("VISA")
    else:
        print("INVALID")



if __name__ == "__main__":
    main()

File: pset1/credit/credit.c
#include <stdio.h>
#include <cs50.h>

int get_starting_num(long card_number, int power);

int main(void)
{
    long card_number;
    do
    {
        card_number = get_long("Number: ");
    }
    while (card_number < 0 || card_number > 9999999999999999);
    // Length of the card, used to get starting num
    int count = 0;
    // Non product digits
    int odd_count = 0;
    // Product digits
    int even_count = 0;
    // Copy of card number to be iterated over
    long num = card_number;
    // Calculates the sum of the product and non-product digits
    while (num > 0)
    {
        count ++;
        int digit = num % 10;
        num = num / 10;

        if (count % 2 == 0)
        {
            even_count = even_count + ((2 * digit) % 10);
            even_count = even_count + ((2 * digit) / 10);
        }
        else
        {
            odd_count += digit;
        }
    }

    // The value of the lugn algo
    int luhn_val = (odd_count + even_count) % 10;

    int starting_digits = get_starting_num(card_number, (count - 2));

    if (luhn_val == 0)
    {
        if ((starting_digits == 34 || starting_digits == 37) && count == 15)
        {
            printf("AMEX\n");
        }
        else if ((starting_digits == 51 || starting_digits == 52 || starting_digits == 53 || starting_digits == 54
                    || starting_digits == 55) && count == 16)
        {
            printf("MASTERCARD\n");
        }
        else
        {
            if (((starting_digits / 10) == 4) && (count == 13 || count == 16))
            {
                printf("VISA\n");
            }
            else
            {
                printf("INVALID\n");
            }
        }
    }
    else
    {
        printf("INVALID\n");
    }
}

// Gets the starting 2 digits of the card number
int get_starting_num(long card_number, int power)
{
    for (int i = 0; i < power; i++)
    {
        card_number = card_number / 10;
    }
    return card_number;
}